### PR TITLE
New `api/label` package, common label set impl

### PR DESCRIPTION
**Pull request by jmacd**
This extracts the `sdk/metric` label set implementation, the `sdk/export/metric` default label encoder, and moves them into `api/label`.  The new type is `*label.Set` supports the same iterator interface, but with the removal of two interfaces `Labels` and `LabelStorage` are no longer needed, since the API uses structs and pointers.

The main support implemented by the SDK's label set was computing a distinct value for every unique set of labels, after de-duplication and last-value-wins semantics are applied. The internal map key used in the SDK is now exposed as `label.Distinct`, which is a struct wrapping a non-exported `interface{}`, wrapping a variable-size array. 

This replaces partly replaces #640 and #641, in that it changes how uniqueness is exposed and resources are represented.

Note that the new `label.Set` type replaces both the metric label set and the resource struct in this PR. In a future PR it should also replace the `correlation.Map` (TODO).

The `label.Set` object manages a cache of encodings by label encoder. This support is expanded to >1 encoding per label set, so that multiple exporters can co-exist without harming performance. The number of encodings is limited to 3 in this PR, but can be expanded in the future.

### CODE REVIEW

File: api/label/set.go [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+	// overwritten values are swapped to the beginning.
+
+	// Check for empty set.
+	if len(kvs) == 0 {
+		return emptySet
+	}
+
+	*tmp = kvs
+
+	// Sort and de-duplicate.  Note: this use of `tmp`
```
Comment (MrAlias) [inline api/label/set.go:266]: Might be better to scope this comment to the sorting only and put the de-duplication near the `for` loop.

```suggestion
	// Stable sort so the following de-duplication can implement last-value-wins semantics.
	// Note: this use of `tmp`
```

File: api/label/set.go [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+	// avoids an allocation because it is a pointer.
+	sort.Stable(tmp)
+
+	*tmp = nil
+
+	position := len(kvs) - 1
+	offset := position - 1
+
+	for ; offset >= 0; offset-- {
```
Comment (MrAlias) [inline api/label/set.go:275]:
```suggestion
	// De-duplicate with last-value-wins semantics. Preserve duplicate values at the beginning of the input slice.
	for ; offset >= 0; offset-- {
```

File: api/label/set.go [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+		position = position - 1
+	}
+
+	return Set{
+		equivalent: computeDistinct(kvs[position:]),
+	}
+}
+
+// computeDistinct returns a `Distinct` using either the fixed- or
+// reflect-oriented code path, depending on the size of the input.
```
Comment (MrAlias) [inline api/label/set.go:289]:
```suggestion
// reflect-oriented code path, depending on the size of the input.
// The input slice is assumed to already be sorted and de-duplicated.
```

File: api/label/set.go [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+	}
+}
+
+// computeDistinctFixed computes a `Distinct` for small slices.  It
+// returns nil if the input is too large for this code path.
+func computeDistinctFixed(kvs []core.KeyValue) interface{} {
+	switch len(kvs) {
+	case 1:
+		ptr := new([1]core.KeyValue)
+		copy((*ptr)[:], kvs)
```
Comment (MrAlias) [inline api/label/set.go:306]: just a nit, but this duplicate code:

```
copy((*ptr)[:], kvs)
return *ptr
```

could be consolidated to follow the `switch` statement.

File: sdk/resource/resource.go [hunk 5]
```diff
@@ -17,95 +17,22 @@
...
 
-// Resource describes an entity about which identifying information and metadata is exposed.
 type Resource struct {
-	sorted []core.KeyValue
-	keySet map[core.Key]struct{}
+	labels label.Set
 }
 
 // New creates a resource from a set of attributes.
 // If there are duplicates keys then the first value of the key is preserved.
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply sdk/resource/resource.go:29]: According to the spec, the Merge semantics are based more on resource precedence:

> The resulting resource MUST have all attributes that are on any of the two input resources. Conflicts (i.e. a key for which attributes exist on both the primary and secondary resource) MUST be handled as follows:
>
> If the value on the primary resource is an empty string, the result has the value of the secondary resource.
> Otherwise, the value of the primary resource is used.

I think if we did a similar thing to the existing `Merge` method and just built a slice with the primary labels being last and then sort/de-dupe that we would implement the spec.

Otherwise, I don't see any issue with the spec on changing the label precedence here. I do wonder if we have logic internal to this codebase that relies on this first-value-wins ordering. I would guess not, but it's just a guess.

File: sdk/resource/resource.go [hunk 6]
```diff
@@ -17,95 +17,22 @@
...
 
 import (
-	"sort"
-	"strings"
-
 	"go.opentelemetry.io/otel/api/core"
+	"go.opentelemetry.io/otel/api/label"
 )
 
-// Resource describes an entity about which identifying information and metadata is exposed.
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply sdk/resource/resource.go:26]: I think the question is specifically about the struct comment that was removed.

File: sdk/resource/resource.go [hunk 7]
```diff
@@ -17,95 +17,22 @@
...
-		}
-
-	}
-	b.WriteRune(')')
-
-	return b.String()
-}
-
-// Attributes returns a copy of attributes from the resource in a sorted order.
-func (r Resource) Attributes() []core.KeyValue {
```
Comment (MrAlias) [inline sdk/resource/resource.go:77]: This rename will need to be propagated to all callers.

File: sdk/trace/trace_test.go [hunk 8]
```diff
@@ -602,7 +602,15 @@ func cmpDiff(x, y interface{}) string {
 	return cmp.Diff(x, y,
 		cmp.AllowUnexported(core.Value{}),
 		cmp.AllowUnexported(export.Event{}),
-		cmp.AllowUnexported(resource.Resource{}))
+		cmp.Comparer(func(a, b *resource.Resource) bool {
+			if a == nil && b == nil {
+				return true
+			} else if a != nil && b != nil {
+				return a.Labels().Equivalent() == b.Labels().Equivalent()
```
Comment (MrAlias) [inline sdk/trace/trace_test.go:609]: Possibly use the convenience method?

```suggestion
				return a.Labels().Equal(b.Labels())
```

File: sdk/metric/sdk.go [hunk 9]
```diff
@@ -144,34 +124,25 @@ type (
 		instrument
 		// recorders maps ordered labels to the pair of
 		// labelset and recorder
-		recorders map[orderedLabels]labeledRecorder
+		recorders map[label.Distinct]labeledRecorder
```
Comment (MrAlias) [inline sdk/metric/sdk.go:127]: :+1:

File: api/label/encoder.go [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+// this choice is arbitrary.  Users will see these in some exporters
+// (e.g., stdout), so the backslash ('\') is used a conventional choice.
+const escapeChar = '\\'
+
+var (
+	_ Encoder = &defaultLabelEncoder{}
+
+	// labelEncoderIDCounter is for generating IDs for other label
+	// encoders.
+	encoderIDCounter int64 = 1
```
[the remark this answers is not in the document]
Comment (krnowak) [reply api/label/encoder.go:67]: The whole point is that zero is invalid. But counter can start at zero, because the first call to Atomic.AddInt64 will return 1, not zero.

File: sdk/metric/batcher/ungrouped/ungrouped.go [hunk 11]
```diff
@@ -28,7 +29,7 @@ type (
 		selector     export.AggregationSelector
 		batchMap     batchMap
 		stateful     bool
-		labelEncoder export.LabelEncoder
+		labelEncoder label.Encoder
```
[the remark this answers is not in the document]
Comment (krnowak) [reply sdk/metric/batcher/ungrouped/ungrouped.go:32]: Yeah, that's fine.

File: api/label/set.go (hunk 4 above)
Comment (krnowak) [reply api/label/set.go:306]: Some of this could be consolidated, but ptr itself is of different type in each case.

File: sdk/resource/resource.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (krnowak) [reply sdk/resource/resource.go:26]: Right.

File: api/label/encoder.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (MrAlias) [reply api/label/encoder.go:67]: > The whole point is that zero is invalid. But counter can start at zero, because the first call to Atomic.AddInt64 will return 1, not zero.

Gotcha :+1:

File: sdk/resource/resource.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jmacd) [reply sdk/resource/resource.go:26]: I literally erased this file and entered these stubs, but now I've restored the former comments and method order. 😁

File: sdk/resource/resource.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jmacd) [reply sdk/resource/resource.go]: I've updated the comment to say that `New` uses last-value-wins, which I believe is an OTel-wide policy, and that `Merge` uses gives the first resource preference.

File: sdk/resource/resource.go (hunk 7 above)
Comment (jmacd) [reply sdk/resource/resource.go:77]: I restored `Attributes`.

File: sdk/trace/trace_test.go (hunk 8 above)
Comment (jmacd) [reply sdk/trace/trace_test.go]: I was able to remove this code from the test.  The problem was when some code is using a pointer and other code is using a struct, that the special support from `cmp` was needed.

I have added a `Set.Equal()` helper method anyway.

File: api/label/set.go [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+// Note that methods are defined `*Set`, although no allocation for
+// `Set` is required.  Callers may avoid memory allocations by:
+//
+// - allocating a `Sortable` for use as a temporary in this method
+// - allocating a `Set` for storing the return value of this
+//   constructor.
+func NewSetWithSortable(kvs []core.KeyValue, tmp *Sortable) Set {
+	// The requitements stated above require that the stable
+	// result be placed in the end of the input slice, while
+	// overwritten values are swapped to the beginning.
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/label/set.go]: Moved, and more comments added.

File: api/label/set.go (hunk 1 above)
Comment (jmacd) [reply api/label/set.go]: Done.

File: api/label/set.go (hunk 2 above)
Comment (jmacd) [reply api/label/set.go:286]: Done.

File: api/label/set.go [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+
+	position := len(kvs) - 1
+	offset := position - 1
+
+	for ; offset >= 0; offset-- {
+		if kvs[offset].Key == kvs[position].Key {
+			continue
+		}
+		kvs[offset], kvs[position-1] = kvs[position-1], kvs[offset]
+		position = position - 1
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/label/set.go]: Done

File: api/label/set.go (hunk 3 above)
Comment (jmacd) [reply api/label/set.go:300]: Done.

File: go.sum [hunk 14]
```diff
@@ -37,6 +37,7 @@ github.com/prometheus/client_model v0.0.0-20190812154241-14fe0d1b01d4/go.mod h1:
 github.com/stretchr/objx v0.1.0/go.mod h1:HFkY916IF+rwdDfMAkV7OtwuqBVzrE8GR6GFx+wExME=
 github.com/stretchr/testify v1.4.0 h1:2E4SXV/wtOkTonXsotYi4li6zVWxYlZuYNCXe9XRJyk=
 github.com/stretchr/testify v1.4.0/go.mod h1:j7eGeouHqKxXV5pUuKE4zz7dFj8WfuZ+81PSLYec5m4=
+go.opentelemetry.io v0.1.0 h1:EANZoRCOP+A3faIlw/iN6YEWoYb1vleZRKm1EvH8T48=
```
[the remark this answers is not in the document]
Comment (jmacd) [reply go.sum]: Not sure how it got there. I ran `go mod tidy`.

File: sdk/metric/correct_test.go [hunk 15]
```diff
@@ -239,28 +240,34 @@ func TestSDKLabelsDeduplication(t *testing.T) {
...
 		actual = append(actual, kvs)
 	}
 
 	require.ElementsMatch(t, allExpect, actual)
 }
 
+func newSet(kvs ...core.KeyValue) *label.Set {
+	labels := &label.Set{}
+	*labels = label.NewSet(kvs...)
+	return labels
```
[the remark this answers is not in the document]
Comment (jmacd) [reply sdk/metric/correct_test.go]: Done.

File: sdk/metric/sdk.go [hunk 16]
```diff
@@ -652,11 +460,16 @@ func (m *SDK) RecordBatch(ctx context.Context, kvs []core.KeyValue, measurements
...
-	var labels labels
+	var labels label.Set
 	for i, meas := range measurements {
 		s := meas.SyncImpl().(*syncInstrument)
 
-		h := s.acquireHandle(kvs, &labels)
+		var labelsPtr *label.Set
+		if i > 0 {
+			labelsPtr = &labels
+		}
```
[the remark this answers is not in the document]
Comment (jmacd) [reply sdk/metric/sdk.go]: This was revised. Note that the former code organization would not have cached label encodings for a batch because it was copying the Set struct. (Lint found this w "copylocks" analysis.)

Therefore, the SDK `record` now contains both a `Label` for the common case and a `*Label` for the batch case. PTAL.

File: api/label/set.go [hunk 17]
```diff
@@ -0,0 +1,372 @@
...
+			return l.encoded[idx]
+		}
+		if !l.encoders[idx].Valid() {
+			l.encoders[idx] = id
+			l.encoded[idx] = r
+			return r
+		}
+	}
+
+	// TODO: This is a performance cliff.  Find a way for this to
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/label/set.go:219]: I'd like to address it with #174.

File: sdk/metric/batcher/ungrouped/ungrouped.go [hunk 18]
```diff
@@ -47,7 +48,7 @@ type (
 var _ export.Batcher = &Batcher{}
 var _ export.CheckpointSet = batchMap{}
 
-func New(selector export.AggregationSelector, labelEncoder export.LabelEncoder, stateful bool) *Batcher {
+func New(selector export.AggregationSelector, labelEncoder label.Encoder, stateful bool) *Batcher {
```
[the remark this answers is not in the document]
Comment (jmacd) [reply sdk/metric/batcher/ungrouped/ungrouped.go:51]: (For another PR, #641).

File: sdk/resource/resource.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jmacd) [reply sdk/resource/resource.go]: > I do wonder if we have logic internal to this codebase that relies on this first-value-wins ordering.

I suspect the answer is no. I found a couple of tests that had to be changed, of course.

File: api/label/set.go [hunk 19]
```diff
@@ -0,0 +1,372 @@
...
+	Set struct {
+		equivalent Distinct
+
+		lock     sync.Mutex
+		encoders [maxConcurrentEncoders]EncoderID
+		encoded  [maxConcurrentEncoders]string
+	}
+
+	// Distinct wraps a variable-size array of `core.KeyValue`,
+	// constructed with keys in sorted order.
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/label/set.go]: Done.

File: api/label/set.go [hunk 20]
```diff
@@ -0,0 +1,372 @@
...
+	Distinct struct {
+		iface interface{}
+	}
+
+	// Iterator allows iterating over the set of labels in order,
+	// sorted by key.
+	Iterator struct {
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/label/set.go]: Done.

File: api/label/set.go [hunk 21]
```diff
@@ -0,0 +1,372 @@
...
+		storage *Set
+		idx     int
+	}
+
+	// Sortable implements `sort.Interface`, used for sorting
+	// `core.KeyValue`.  This is an exported type to support an
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/label/set.go]: Done.

**Comment by jmacd** [comment]
PTAL. Note that this includes as a prerequisite the changes from https://github.com/open-telemetry/opentelemetry-go/pull/654

File: sdk/resource/resource.go [hunk 22]
```diff
@@ -17,101 +17,118 @@
...
-		}
-
+// String implements the Stringer interface and provides a
+// human-readable form of the resource.
+//
+// Avoid using this representation as the key in a map of resources,
+// use Equivalent() as the key instead.
+func (r *Resource) String() string {
+	if r == nil {
+		r = Empty()
```
Comment (MrAlias) [inline sdk/resource/resource.go:53]: nit:
```suggestion
		return ""
```

File: sdk/resource/resource.go [hunk 23]
```diff
@@ -17,101 +17,118 @@
...
+	}
+	if eq == nil {
+		eq = Empty()
+	}
+	return r.Equivalent() == eq.Equivalent()
 }
 
 // Merge creates a new resource by combining resource a and b.
 // If there are common key between resource a and b then value from resource a is preserved.
 // If one of the resources is nil then the other resource is returned without creating a new one.
```
Comment (MrAlias) [inline sdk/resource/resource.go:89]: Don't think this is true anymore, which (it not being true anymore) I also think is appropriate for an immutable Resource.
```suggestion
```

File: sdk/resource/resource.go [hunk 24]
```diff
@@ -17,101 +17,118 @@
...
+	if r == nil {
+		r = Empty()
+	}
+	return r.labels.MarshalJSON()
+}
+
+// Len returns the number of unique key-values in this Resource.
+func (r *Resource) Len() int {
+	if r == nil {
+		r = Empty()
```
Comment (MrAlias) [inline sdk/resource/resource.go:131]: nit:

```suggestion
		return 0
```

Minor, but would avoid the possible use of a reflect value. :shrug:

File: sdk/resource/iterator_test.go [hunk 25]
```diff
@@ -12,21 +12,21 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package resource
```
Comment (MrAlias) [inline sdk/resource/iterator_test.go:15]: Could probably just remove this file now that there is a unified place (not here) that implements (and tests) the iterator. These are just testing that implementation anyways.

File: exporters/otlp/otlp_metric_test.go [hunk 26]
```diff
@@ -681,6 +682,9 @@ func runMetricExportTest(t *testing.T, exp *Exporter, rs []record, expected []me
 				continue
 			}
 			assert.ElementsMatch(t, ilm.GetMetrics(), g, "metrics did not match for:\n\tResource: %s\n\tInstrumentationLibrary: %s\n", k.resource, k.instrumentationLibrary)
+			// if diff := cmp.Diff(ilm.GetMetrics(), g); diff != "" {
```
Comment (MrAlias) [inline exporters/otlp/otlp_metric_test.go:685]: I was kind of confused when I saw this added because I had written almost exactly the same thing when I originally wrote the tests. Maybe we should just switch to `cmd.Diff` instead of `assert.ElementsMatch` permanently?

File: sdk/metric/correct_test.go [hunk 27]
```diff
@@ -241,28 +242,33 @@ func TestSDKLabelsDeduplication(t *testing.T) {
...
 
-		kvs := export.IteratorToSlice(rec.Labels().Iter())
+		kvs := rec.Labels().ToSlice()
 		actual = append(actual, kvs)
 	}
 
 	require.ElementsMatch(t, allExpect, actual)
 }
 
+func newSet(kvs ...core.KeyValue) *label.Set {
```
Comment (MrAlias) [inline sdk/metric/correct_test.go:252]: nit: might want to change to `newSetIter` and return a `label.Iterator` given all calls use the returned `Iter()`.

File: exporters/otlp/internal/transform/span.go [hunk 28]
```diff
@@ -32,14 +33,14 @@ func SpanData(sdl []*export.SpanData) []*tracepb.ResourceSpans {
...
-	// Group by the unique string representation of the Resource.
-	rsm := make(map[string]*tracepb.ResourceSpans)
+	// Group by the distinct representation of the Resource.
+	rsm := make(map[label.Distinct]*tracepb.ResourceSpans)
 
 	for _, sd := range sdl {
 		if sd != nil {
-			var key string
+			var key label.Distinct
 			if sd.Resource != nil {
```
Comment (MrAlias) [inline exporters/otlp/internal/transform/span.go:42]: These guards aren't necessary now that `Resrource.Equivalent()` handles `nil` correctly.

File: api/label/iterator.go [hunk 29]
```diff
@@ -0,0 +1,77 @@
...
+
+// Label returns current core.KeyValue. Must be called only after Next returns
+// true.
+func (i *Iterator) Label() core.KeyValue {
+	kv, _ := i.storage.Get(i.idx)
+	return kv
+}
+
+// Attribute is a synonym for Label().
+func (i *Iterator) Attribute() core.KeyValue {
```
Comment (MrAlias) [inline api/label/iterator.go:43]: Not an issue for this PR (or project), but it makes it pretty clear OpenTelemetry is a pieced together project when things like this exist :facepalm:.  Hopefully one day this can be unified in the spec.

File: sdk/resource/resource.go (hunk 24 above)
Comment (jmacd) [reply sdk/resource/resource.go:131]: Done. (There's no reflection in `Empty()`--not sure what you mean.)

File: exporters/otlp/otlp_metric_test.go (hunk 26 above)
Comment (jmacd) [reply exporters/otlp/otlp_metric_test.go:685]: I forgot about that. I think we should leave it in. I couldn't read the output prior to adding this.

Comment (jmacd) [reply exporters/otlp/otlp_metric_test.go:685]: Oh! I remember why, now. I wanted to leave that in, but it's testing more than we want. It's an ordered test, unlike `ElementsMatch`. I can't figure out how to make an elements-match-independent-of-order test out of `gocmp`.

File: sdk/resource/resource.go (hunk 24 above)
Comment (MrAlias) [reply sdk/resource/resource.go]: Ah, your are correct. I was mis-remembering to when unique values were stored with reflect Values and confusing the two. My bad.

File: sdk/resource/resource.go (hunk 23 above)
Comment (jmacd) [reply sdk/resource/resource.go]: Done.

File: sdk/resource/iterator_test.go (hunk 25 above)
Comment (jmacd) [reply sdk/resource/iterator_test.go:15]: I moved this test into the api/label package.

File: exporters/otlp/internal/transform/span.go (hunk 28 above)
Comment (jmacd) [reply exporters/otlp/internal/transform/span.go]: Done.

File: api/label/iterator.go (hunk 29 above)
Comment (jmacd) [reply api/label/iterator.go:43]: Yesss. 😀

File: sdk/resource/resource.go (hunk 22 above)
Comment (jmacd) [reply sdk/resource/resource.go]: Done.

File: sdk/metric/correct_test.go (hunk 27 above)
Comment (jmacd) [reply sdk/metric/correct_test.go]: Done.

**Comment by jmacd** [comment]
(I added more tests.)

File: sdk/metric/sdk.go [hunk 30]
```diff
@@ -117,15 +97,22 @@ type (
 		// supports checking for no updates during a round.
 		collectedCount int64
 
-		// labels is the processed label set for this record.
+		// storage is the stored label set for this record,
+		// except in cases where a label set is shared due to
+		// batch recording.
 		//
 		// labels has to be aligned for 64-bit atomic operations.
```
Comment (krnowak) [inline sdk/metric/sdk.go:104]: This could be dropped, it's out of date.
```suggestion
```

File: sdk/metric/sdk.go [hunk 31]
```diff
@@ -117,15 +97,22 @@ type (
...
-		labels labels
+		storage label.Set
+
+		// labels is the processed label set for this record.
+		// this may refer to the `storage` field if this label
+		// set is not shared.
```
Comment (krnowak) [inline sdk/metric/sdk.go:109]: This may also refer to the `storage` field when this label set _is_ shared, but this record was the first in batch. I'd rephrase it like so:
```suggestion
		// This may refer to the `storage` field in another record
		// if this label set is shared.
```

File: api/label/set.go [hunk 32]
```diff
@@ -0,0 +1,388 @@
...
+// Iter returns an iterator for visiting the labels in this set.
+func (l *Set) Iter() Iterator {
+	return Iterator{
+		storage: l,
+		idx:     -1,
+	}
+}
+
+// Return the set of labels belonging to this set, sorted, where keys
+// appear no more than once.
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/label/set.go]: Done.

File: api/label/set.go [hunk 33]
```diff
@@ -0,0 +1,388 @@
...
+
+// NewSetWithSortable returns a new `*Set`.
+//
+// Duplicate keys are eliminated by taking the last value.  This
+// re-orders the input slice so that unique last-values are contiguous
+// at the end of the slice.
+//
+// This ensures the following:
+//
+// - Last-value wins semantics
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/label/set.go]: Done.

File: api/label/set.go [hunk 34]
```diff
@@ -0,0 +1,388 @@
...
+	// Stable sort so the following de-duplication can implement
+	// last-value-wins semantics.
+	sort.Stable(tmp)
+
+	*tmp = nil
+
+	position := len(kvs) - 1
+	offset := position - 1
+
+	// The requitements stated above require that the stable
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/label/set.go]: Done.

File: api/label/set.go [hunk 35]
```diff
@@ -0,0 +1,388 @@
...
+// works for any size input.
+func computeDistinctReflect(kvs []core.KeyValue) interface{} {
+	at := reflect.New(reflect.ArrayOf(len(kvs), keyValueType)).Elem()
+	for i, kv := range kvs {
+		*(at.Index(i).Addr().Interface().(*core.KeyValue)) = kv
+	}
+	return at.Interface()
+}
+
+func (l *Set) MarshalJSON() ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/label/set.go]: Done.

File: api/label/set_test.go [hunk 36]
```diff
@@ -0,0 +1,105 @@
...
+		}
+		// No other Distinct values are equal to this.
+		for _, strings := range d2s {
+			if strings[0] == s {
+				continue
+			}
+			for _, otherString := range strings {
+				require.NotEqual(t, otherString, s)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/label/set_test.go]: Thank you! I used this logic, same below.

File: api/label/set_test.go [hunk 37]
```diff
@@ -0,0 +1,105 @@
...
+		// No other String values are equal to this.
+		for _, distinct := range s2d {
+			if distinct[0] == d {
+				continue
+			}
+			for _, otherDistinct := range distinct {
+				require.NotEqual(t, otherDistinct, d)
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/label/set_test.go]: Done.

File: exporters/trace/jaeger/jaeger.go [hunk 38]
```diff
@@ -204,12 +204,13 @@ func spanDataToThrift(data *export.SpanData) *gen.Span {
 		}
 	}
 
-	// TODO (rghetia): what to do if a resource key is the same as one of the attribute's key
-	// TODO (rghetia): is there a need for prefixing keys with "resource-"?
+	// TODO (jmacd): OTel has a broad "last value wins"
+	// semantic. Should resources be appended before span
+	// attributes, above, to allow span attributes to ovewrite
```
[the remark this answers is not in the document]
Comment (jmacd) [reply exporters/trace/jaeger/jaeger.go]: Done.
